crates/storage/provider/src/providers/rocksdb/invariants.rs

```diff
@@ -0,0 +1,413 @@
+    {
+        let mut unwind_target: Option<BlockNumber> = None;
+
+        // Check TransactionHashNumbers if stored in RocksDB
+        if provider.cached_storage_settings().transaction_hash_numbers_in_rocksdb &&
+            let Some(target) = self.check_transaction_hash_numbers(provider)?
+        {
+            unwind_target = Some(unwind_target.map_or(target, |t| t.min(target)));
+        }
+
+        // Check StoragesHistory if stored in RocksDB
+        if provider.cached_storage_settings().storages_history_in_rocksdb &&
+            let Some(target) = self.check_storages_history(provider)?
+        {
+            unwind_target = Some(unwind_target.map_or(target, |t| t.min(target)));
+        }
+
+        Ok(unwind_target)
+    }
+
+    /// Checks invariants for the `TransactionHashNumbers` table.
+    ///
+    /// Returns a block number to unwind to if `RocksDB` is behind the checkpoint.
+    fn check_transaction_hash_numbers<Provider>(
+        &self,
+        provider: &Provider,
+    ) -> ProviderResult<Option<BlockNumber>>
+    where
+        Provider: DBProvider + StageCheckpointReader,
+    {
+        use reth_db::cursor::DbCursorRO;
+        use reth_db_api::transaction::DbTx;
+
+        // Get the TransactionLookup stage checkpoint
+        let checkpoint = provider
+            .get_stage_checkpoint(StageId::TransactionLookup)?
+            .map(|cp| cp.block_number)
+            .unwrap_or(0);
+
+        // Get the last entry from RocksDB TransactionHashNumbers
+        let rocks_last = self.last::<tables::TransactionHashNumbers>()?;
+
+        match rocks_last {
+            Some((_hash, tx_number)) => {
+                // If checkpoint is 0 but we have data, that's inconsistent
+                if checkpoint == 0 && tx_number > 0 {
+                    // RocksDB has data but stage hasn't run - need to clear
+                    return Ok(Some(0));
+                }
+
+                // Map tx_number to block_number using TransactionBlocks table
+                // TransactionBlocks key is the highest tx ID in each block
+                let mut cursor = provider.tx_ref().cursor_read::<tables::TransactionBlocks>()?;
+
+                // Seek to the first entry where key >= tx_number
+                if let Some((_, rocks_block)) = cursor.seek(tx_number)? {
+                    // rocks_block is the block containing this transaction
+                    // If checkpoint block > rocks_block, RocksDB is behind
+                    if checkpoint > rocks_block {
+                        tracing::info!(
+                            target: "reth::providers::rocksdb",
+                            rocks_max_tx = tx_number,
+                            rocks_block,
+                            checkpoint,
+                            "TransactionHashNumbers behind checkpoint, unwind needed"
+                        );
+                        return Ok(Some(rocks_block));
+                    }
```

Comment:
either here or somewhere else, we should cover the other way around which is the most common occurrence
`else if checkpoint < rocks_block { self.delete.... } ` 

when appending data, mdbx stage checkpoint will be the last thing to be committed to disk :
https://github.com/paradigmxyz/reth/blob/64909d33e6b7ab60774e37f5508fb5ad17f41897/crates/storage/provider/src/providers/database/provider.rs#L3176-L3178

when unwinding storage (eg pipeline unwind/engine deep reorg), mdbx stage checkpoint will be the first thing to be commited to disk:
https://github.com/paradigmxyz/reth/blob/64909d33e6b7ab60774e37f5508fb5ad17f41897/crates/storage/provider/src/providers/database/provider.rs#L3169-L3175

Reply:
> checkpoint > rocks_block

for this to happen, something has gone terrible wrong if we follow the above rules

Reply:
good point, added changes to remove excess rocksdb entries when it is ahead of mdbx, ptal